File: python/src/ancilis/middleware/middleware.py
"""MCP middleware — intercepts tool calls, evaluates via engine, enforces decisions."""

from __future__ import annotations

import logging
from pathlib import Path
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from mcp import ClientSession
    from mcp.types import CallToolResult
else:
    ClientSession = Any
    CallToolResult = Any

try:
    from mcp.types import TextContent
except ImportError:
    class TextContent:  # type: ignore[no-redef]
        """Fallback type when MCP is not installed."""

        pass

from ancilis.config import ResolvedConfig, load_config
from ancilis.engine.engine import Engine
from ancilis.engine.registry import ToolEntry, ToolRegistry, ToolStatus
from ancilis.engine.result import EvaluationResult
from ancilis.middleware.action_builder import build_action
from ancilis.middleware.discovery import DriftEvent, register_tools_from_list
from ancilis.evidence.store import EvidenceStore
from ancilis.middleware.response_scanner import ScanResult, scan_response

logger = logging.getLogger("ancilis.middleware")


class BlockedToolCallError(Exception):
    """Raised when a tool call is blocked by policy enforcement."""

    def __init__(self, tool_name: str, evaluation: EvaluationResult):
        self.tool_name = tool_name
        self.evaluation = evaluation
        # Build developer-facing message using display names
        reasons: list[str] = []
        for r in evaluation.control_results:
            if r.result in ("FAIL", "ERROR"):
                display = r.display_name or r.control_name
                reasons.append(display.lower())
        reason_str = ", ".join(reasons) if reasons else "policy violation"
        self.display_message = (
            f"Ancilis [blocked]: Tool '{tool_name}' blocked — {reason_str}.\n"
            f"  To approve: ancilis approve-tool {tool_name}\n"
            f"  To review: ancilis status"
        )
        super().__init__(
            f"Tool call '{tool_name}' blocked by policy. "
            f"Failed controls: {', '.join(r.control_id for r in evaluation.control_results if r.result in ('FAIL', 'ERROR'))}"
        )


class AncilisMiddleware:
    """Wraps an MCP ClientSession to intercept and evaluate tool calls."""

    def __init__(
        self,
        session: ClientSession,
        config_path: str | Path | None = None,
        config: ResolvedConfig | None = None,
        evidence_store: EvidenceStore | None = None,
    ) -> None:
        if config is not None:
            self._config = config
        elif config_path is not None:
            self._config = load_config(path=config_path)
        else:
            self._config = load_config(raw={"agent": {"name": "ancilis-agent"}})

        self._session = session
        self._registry = ToolRegistry()

        # Pre-seed registry with config-approved tools so PR-03 passes
        # even before list_tools() / discovery runs.
        for tool_name in self._config.tools_allowed:
            self._registry.register(ToolEntry(
                name=tool_name,
                status=ToolStatus.APPROVED,
            ))

        self._engine = Engine(self._config, registry=self._registry)

        self._evaluation_log: list[EvaluationResult] = []
        self._scan_results: list[ScanResult] = []
        self._drift_events: list[DriftEvent] = []
        self._evidence_store = evidence_store if evidence_store is not None else EvidenceStore(self._config)
        self._issue_count: int = 0
        self._closed: bool = False

    @property
    def evidence_store(self) -> EvidenceStore:
        return self._evidence_store

    @property
    def config(self) -> ResolvedConfig:
        return self._config

    @property
    def registry(self) -> ToolRegistry:
        return self._registry

    @property
    def evaluation_log(self) -> list[EvaluationResult]:
        return list(self._evaluation_log)

    @property
    def scan_results(self) -> list[ScanResult]:
        return list(self._scan_results)

    @property
    def drift_events(self) -> list[DriftEvent]:
        return list(self._drift_events)

    async def call_tool(
        self,
        name: str,
        arguments: dict[str, Any] | None = None,
    ) -> CallToolResult:
        """Intercept a tool call, evaluate, enforce, forward or block."""
        # 1. Build Action object
        action = build_action(name, arguments, self._config, self._registry)

        # 2. Evaluate via engine
        evaluation = self._engine.evaluate(action)
        self._evaluation_log.append(evaluation)

        # 3. Store evidence
        self._evidence_store.store(evaluation, tool_name=name)

        # Track issues for summary line
        has_issues = any(r.result in ("FAIL", "ERROR") for r in evaluation.control_results)
        if has_issues:
            self._issue_count += 1

        # 4. Enforce decision
        if evaluation.decision == "BLOCK":
            error = BlockedToolCallError(name, evaluation)
            logger.warning("%s", error.display_message)
            raise error

        # 5. Forward to MCP server
        try:
            result = await self._session.call_tool(name, arguments)
        except Exception:
            logger.exception("MCP server error calling tool '%s'", name)
            raise

        # 6. Scan response
        response_text = self._extract_response_text(result)
        if response_text:
            scan = scan_response(name, response_text)
            if scan.patterns or scan.encryption_findings:
                self._scan_results.append(scan)
                # Alert-level: sensitive data detected
                if scan.patterns:
                    pattern_names = [p.pattern_type for p in scan.patterns]
                    logger.warning(
                        "Ancilis [alert]: Sensitive data pattern (%s) detected in "
                        "outbound parameters for tool '%s'.\n"
                        "  Recommended: review data handling declarations in ancilis.yaml\n"
                        "  Details: ancilis status",
                        ", ".join(pattern_names), name,
                    )
                for finding in scan.encryption_findings:
                    logger.info("Positive finding: %s", finding.detail)

        # 7. Return result to agent
        return result

    async def list_tools(self) -> Any:
        """Forward list_tools to MCP server and auto-register discovered tools."""
        result = await self._session.list_tools()
        tools = result.tools if hasattr(result, "tools") else []

        drift = register_tools_from_list(
            tools, self._registry, pre_approved=self._config.tools_allowed
        )
        self._drift_events.extend(drift)

        registered_count = len(tools)
        logger.info("Auto-discovered %d tools", registered_count)

        return result

    def get_recommendations(self) -> list[str]:
        """Get all accumulated classification recommendations."""
        recs: list[str] = []
        for scan in self._scan_results:
            recs.extend(scan.recommendations)
        return recs

    def get_summary_line(self) -> str:
        """Get the single-line summary of middleware activity.

        This is the default console output — one line, periodic.
        """
        total = len(self._evaluation_log)
        issues = self._issue_count
        return f"Ancilis: {total} tool calls evaluated. {issues} issues. Run `ancilis status` for details."

    def get_last_evaluation(self) -> EvaluationResult | None:
        """Get the most recent evaluation result."""
        return self._evaluation_log[-1] if self._evaluation_log else None

    def close(self) -> None:
        """Flush and close the evidence store. Safe to call multiple times."""
        if not self._closed:
            self._evidence_store.close()
            self._closed = True

    async def __aenter__(self) -> AncilisMiddleware:
        return self

    async def __aexit__(self, *exc: Any) -> None:
        self.close()

    @staticmethod
    def _extract_response_text(result: CallToolResult) -> str:
        """Extract text content from an MCP CallToolResult."""
        parts: list[str] = []
        for content in result.content:
            if isinstance(content, TextContent):
                parts.append(content.text)
            elif hasattr(content, "text"):
                parts.append(str(content.text))
        return "\n".join(parts)

File: typescript/src/ancilis/middleware/middleware.ts
/** MCP middleware — intercepts tool calls, evaluates via engine, enforces decisions. */

import type { ResolvedConfig } from "../config/index.js";
import { loadConfig } from "../config/index.js";
import { Engine } from "../engine/engine.js";
import { ToolRegistry } from "../engine/registry.js";
import type { EvaluationResult } from "../engine/result.js";
import { buildAction } from "./action-builder.js";
import { registerToolsFromList } from "./discovery.js";
import type { DriftEvent } from "./discovery.js";
import { EvidenceStore } from "../evidence/store.js";
import { scanResponse } from "./response-scanner.js";
import type { ScanResult } from "./response-scanner.js";

/** Minimal interface for an MCP client — avoids tight coupling to SDK version. */
export interface McpClientLike {
  callTool(params: { name: string; arguments?: Record<string, unknown> }, ...rest: unknown[]): Promise<{ content: Array<{ type: string; text?: string; [key: string]: unknown }>; [key: string]: unknown }>;
  listTools(...args: unknown[]): Promise<{ tools: Array<{ name: string; description?: string; [key: string]: unknown }>; [key: string]: unknown }>;
}

export class BlockedToolCallError extends Error {
  tool_name: string;
  evaluation: EvaluationResult;
  displayMessage: string;

  constructor(toolName: string, evaluation: EvaluationResult) {
    const failed = evaluation.controlResults
      .filter(r => r.result === "FAIL" || r.result === "ERROR")
      .map(r => r.controlId);
    super(`Tool call '${toolName}' blocked by policy. Failed controls: ${failed.join(", ")}`);
    this.name = "BlockedToolCallError";
    this.tool_name = toolName;
    this.evaluation = evaluation;
    // Developer-facing message using display names
    const reasons = evaluation.controlResults
      .filter(r => r.result === "FAIL" || r.result === "ERROR")
      .map(r => (r.displayName ?? r.controlName).toLowerCase());
    const reasonStr = reasons.length > 0 ? reasons.join(", ") : "policy violation";
    this.displayMessage =
      `Ancilis [blocked]: Tool '${toolName}' blocked — ${reasonStr}.\n` +
      `  To approve: ancilis approve-tool ${toolName}\n` +
      `  To review: ancilis status`;
  }
}

export interface AncilisMiddlewareOptions {
  configPath?: string;
  config?: ResolvedConfig;
}

export class AncilisMiddleware {
  private _config: ResolvedConfig;
  private _client: McpClientLike;
  private _registry: ToolRegistry;
  private _engine: Engine;
  private _evaluationLog: EvaluationResult[] = [];
  private _scanResults: ScanResult[] = [];
  private _driftEvents: DriftEvent[] = [];
  private _evidenceStore: EvidenceStore;
  private _issueCount = 0;

  constructor(client: McpClientLike, options: AncilisMiddlewareOptions = {}) {
    if (options.config) {
      this._config = options.config;
    } else if (options.configPath) {
      this._config = loadConfig({ path: options.configPath });
    } else {
      this._config = loadConfig({ raw: { agent: { name: "ancilis-agent" } } });
    }

    this._client = client;
    this._registry = new ToolRegistry();
    this._engine = new Engine(this._config, { registry: this._registry });
    this._evidenceStore = new EvidenceStore(this._config);
  }

  get config(): ResolvedConfig { return this._config; }
  get registry(): ToolRegistry { return this._registry; }
  get evidenceStore(): EvidenceStore { return this._evidenceStore; }
  get evaluationLog(): EvaluationResult[] { return [...this._evaluationLog]; }
  get scanResults(): ScanResult[] { return [...this._scanResults]; }
  get driftEvents(): DriftEvent[] { return [...this._driftEvents]; }

  async callTool(name: string, args?: Record<string, unknown>): Promise<{ content: Array<{ type: string; text?: string; [key: string]: unknown }>; [key: string]: unknown }> {
    // 1. Build Action
    const action = buildAction(name, args, this._config, this._registry);

    // 2. Evaluate
    const evaluation = this._engine.evaluate(action);
    this._evaluationLog.push(evaluation);

    // 3. Store evidence
    await this._evidenceStore.store(evaluation, name);

    // Track issues for summary line
    if (evaluation.controlResults.some(r => r.result === "FAIL" || r.result === "ERROR")) {
      this._issueCount++;
    }

    // 4. Enforce
    if (evaluation.decision === "BLOCK") {
      throw new BlockedToolCallError(name, evaluation);
    }

    // 5. Forward to MCP server
    const result = await this._client.callTool({ name, arguments: args });

    // 6. Scan response
    const responseText = this.extractResponseText(result);
    if (responseText) {
      const scan = scanResponse(name, responseText);
      if (scan.patterns.length > 0 || scan.encryptionFindings.length > 0) {
        this._scanResults.push(scan);
      }
    }

    // 7. Return
    return result;
  }

  async listTools(): Promise<{ tools: Array<{ name: string; description?: string; [key: string]: unknown }>; [key: string]: unknown }> {
    const result = await this._client.listTools();
    const tools = result.tools ?? [];

    const drift = registerToolsFromList(
      tools.map(t => ({ name: t.name, description: t.description })),
      this._registry,
    );
    this._driftEvents.push(...drift);

    return result;
  }

  getRecommendations(): string[] {
    const recs: string[] = [];
    for (const scan of this._scanResults) {
      recs.push(...scan.recommendations);
    }
    return recs;
  }

  getSummaryLine(): string {
    const total = this._evaluationLog.length;
    const issues = this._issueCount;
    return `Ancilis: ${total} tool calls evaluated. ${issues} issues. Run \`ancilis status\` for details.`;
  }

  getLastEvaluation(): EvaluationResult | undefined {
    return this._evaluationLog[this._evaluationLog.length - 1];
  }

  private extractResponseText(result: { content: Array<{ type: string; text?: string; [key: string]: unknown }>; [key: string]: unknown }): string {
    const parts: string[] = [];
    if (Array.isArray(result.content)) {
      for (const item of result.content) {
        if (typeof item === "object" && item !== null && "text" in item && typeof item.text === "string") {
          parts.push(item.text);
        }
      }
    }
    return parts.join("\n");
  }
}
